op-bindings/etherscan/client.go

```diff
@@ -0,0 +1,368 @@
+// behavior when dealing with blockchain data from unsupported networks.
+//
+// Parameters:
+// - chainId: An integer representing the chain ID to be validated.
+//
+// The function returns an error in the following scenario:
+// - If the chain ID is not one of the supported IDs.
+func validateChainId(chainId int) error {
+	switch chainId {
+	case 1, 10:
+	default:
+		return fmt.Errorf("unsupported source chain ID: %d", chainId)
+	}
+
+	return nil
+}
+
+// NewClient initializes a new Client struct with the provided parameters. This client is used to interact with the Etherscan API for both the source and compare chains.
+//
+// Parameters:
+// - sourceChainId: An integer representing the chain ID of the source chain.
+// - compareChainId: An integer representing the chain ID of the compare chain.
+// - sourceApiKey: A string representing the Etherscan API key for the source chain.
+// - compareApiKey: A string representing the Etherscan API key for the compare chain (if needed).
+// - apiMaxRetries: An integer representing the maximum number of retries for an API call.
+// - apiRetryDelay: An integer representing the delay (in seconds) between API call retries.
+//
+// The function returns a pointer to a Client struct and an error in the following scenarios:
+// - If either the source or compare API key is missing when required.
+// - If either the source or compare chain ID is not supported.
+func NewClient(
+	sourceChainId,
+	compareChainId int,
+	sourceApiKey,
+	compareApiKey string,
+	apiMaxRetries,
+	apiRetryDelay int,
+) (*Client, error) {
+	if sourceApiKey == "" {
+		return nil, errors.New("an Etherscan API key for the source chain is required, but none was provided")
+	}
+
+	if compareChainId > 0 && compareApiKey == "" {
+		return nil, errors.New("an Etherscan API key for the compare chain is required, but none was provided")
+	}
+
+	if err := validateChainId(sourceChainId); err != nil {
+		return nil, err
+	}
+	if err := validateChainId(compareChainId); err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		apiMaxRetries: apiMaxRetries,
+		apiRetryDelay: time.Duration(apiRetryDelay) * time.Second,
+		baseUrls: map[int]string{
+			1:  "https://api.etherscan.io/api",
+			10: "https://api-optimistic.etherscan.io/api",
+		},
+		apiKeys: map[int]string{
+			sourceChainId:  sourceApiKey,
+			compareChainId: compareApiKey,
+		},
+	}, nil
+}
+
+// fetch performs an HTTP GET request to the specified URL and retrieves the response body.
+//
+// Parameters:
+// - url: A string representing the URL to which the HTTP GET request is sent.
+//
+// The function returns a byte slice and an error in the following scenarios:
+// - The byte slice contains the response body if the HTTP request is successful.
+// - An error is returned if any issues occur during the HTTP request or while reading the response body.
+func (client *Client) fetch(url string) ([]byte, error) {
+	resp, err := http.Get(url)
```

Comment:
maybe error here if the response status code is not `http.StatusOK`? Not sure what the status code is when etherscan report an error that you want to handle

Reply:
I'd rather the calling function handle the error because there's the Etherscan API and RPC proxy API that return different responses/errors